File: src/main/go/leetcode/lists/hot100/LeetCode22.go
package hot100

func generateParenthesis(n int) []string {
	if n == 0 {
		return []string{}
	}
	var res []string
	generateParenthesisHelper(&res, 0, 0, n, "")
	return res
}

func generateParenthesisHelper(res *[]string, left, right, n int, s string) {
	if len(s) == n*2 {
		*res = append(*res, s)
		return
	}
	if left < n {
		generateParenthesisHelper(res, left+1, right, n, s+"(")
	}
	if right < left {
		generateParenthesisHelper(res, left, right+1, n, s+")")
	}
}

File: src/main/java/leetcode/lists/topinterview/LeetCode22.java
package leetcode.lists.topinterview;

import java.util.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/21 01:05
 */

public class LeetCode22 {

    public List<String> generateParenthesis(int n) {
        if (n == 0) {
            return new ArrayList<>();
        }
        List<String> list = new ArrayList<>();
        dfs(0, 0, n, list, new StringBuilder());
        return list;
    }

    private void dfs(int left, int right, int n, List<String> list, StringBuilder sb) {

        if (sb.length() == n * 2) {
            list.add(sb.toString());
            return;
        }

        if (left < n) {
            sb.append("(");
            dfs(left + 1, right, n, list, sb);
            sb.deleteCharAt(sb.length() - 1);
        }

        if (right < left) {
            sb.append(")");
            dfs(left, right + 1, n, list, sb);
            sb.deleteCharAt(sb.length() - 1);
        }
    }

}


